src/sync.rs

```diff
@@ -74,58 +75,137 @@ impl SyncSession {
+        let mut buf_out = vec![0u8; BUFFER_SIZE];
+
+        if let Ok((size, src_addr)) = socket.recv_from(&mut buf_out[..]) {
+            unsafe {
+                buf_out.set_len(size);
+            }
+            Ok(ResponseItemInt {
+                address: src_addr.ip().to_string(),
+                data: buf_out,
+            })
+        } else {
+            Err(SnmpError::ReceiveError)
+        }
+    }
+
     pub fn get<T>(&mut self, names: &[T]) -> SnmpResult<SnmpPdu>
     where
         T: AsRef<[u32]>,
     {
         let req_id = self.req_id.0;
-        pdu::build_get(self.community.as_slice(), req_id, names, &mut self.send_pdu)?;
-        let recv_len = Self::send_and_recv(&self.socket, &self.send_pdu, &mut self.recv_buf[..])?;
         self.req_id += Wrapping(1);
-        let pdu_bytes = &self.recv_buf[..recv_len];
-        handle_response(req_id, self.community.as_slice(), pdu_bytes.into())
+
+        pdu::build_get(self.community.as_slice(), req_id, names, &mut self.send_pdu)?;
+
+        let response = self.send_and_recv(&self.send_pdu)?;
+
+        handle_response(req_id, self.community.as_slice(), response.data.as_slice())
     }
 
     pub fn getnext(&mut self, name: &[u32]) -> SnmpResult<SnmpPdu> {
         let req_id = self.req_id.0;
+        self.req_id += Wrapping(1);
+
         pdu::build_getnext(self.community.as_slice(), req_id, name, &mut self.send_pdu)?;
-        let recv_len = Self::send_and_recv(&self.socket, &self.send_pdu, &mut self.recv_buf[..])?;
+
+        let response = self.send_and_recv(&self.send_pdu)?;
+
+        handle_response(req_id, self.community.as_slice(), response.data.as_slice())
+    }
+
+    pub fn get_all_responses<T>(
+        &mut self,
+        names: &[T],
+        timeout: Duration,
+    ) -> SnmpResult<Vec<ResponseItem>>
+    where
+        T: AsRef<[u32]>,
+    {
+        let req_id = self.req_id.0;
         self.req_id += Wrapping(1);
-        let pdu_bytes = &self.recv_buf[..recv_len];
-        handle_response(req_id, self.community.as_slice(), pdu_bytes.into())
+
+        let socket = Self::create_socket(self.destination, Some(timeout), true)
+            .map_err(|_| SnmpError::SocketError)?;
+
+        // send
+        pdu::build_get(self.community.as_slice(), req_id, names, &mut self.send_pdu)?;
+        socket
+            .send_to(&self.send_pdu[..], self.destination)
+            .map_err(|_| SnmpError::SendError)?;
+
+        let ts1 = Instant::now();
+
+        // recv all responses
+        let mut vec1: Vec<ResponseItemInt> = Vec::new();
+        loop {
+            let response = Self::recv_one(&socket);
+            if response.is_err() || ts1.elapsed() >= timeout {
+                // skip any errors or delayed response
+                break;
+            }
+            vec1.push(response.unwrap());
+        }
+
+        // parsing to SnmpPdu
+        let mut vec2: Vec<ResponseItem> = Vec::new();
+        for item in vec1.iter() {
```

Comment:
`.iter()` is not needed here, Vec implements `IntoIterator` trait and can be used directly with for loop. If you want non-consuming for a reference can be taken (e.g. `for item in &vec1`).

Reply:
`for item in vec.iter()` is the same as `for item in &vec`. It is non-consuming form that produces references.
If you use consuming form you will get values, not references, but the values are not copied, they are moved. From efficiency perspective it is the same.
The difference is that with non-consuming iteration you can still use this vector afterwards. With consuming operation you can't, the vector is "consumed".